File: src/bobbin/menus/MainMenu.java
package bobbin.menus;

import bobbin.boundaries.Room;
import bobbin.characters.GameCharacter;
import bobbin.characters.PlayerCharacter;
import bobbin.constants.Actions;
import bobbin.constants.Items;
import bobbin.interaction.ConsolePrompt;
import bobbin.interaction.ExitToException;
import bobbin.interaction.Printers;
import bobbin.interaction.actions.ActionList;
import bobbin.interaction.actions.BaseAction;
import bobbin.interaction.console.Console;
import bobbin.io.gamedata.SaveGameSerial;
import bobbin.io.settings.Settings;
import bobbin.items.BaseGameEntity;

import java.io.File;
import java.io.IOException;

public class MainMenu extends Menu {

    static class DummyPlayerCharacter extends PlayerCharacter {
        public DummyPlayerCharacter(String name, String description, Room location) {
            super(name, description, location);
        }
    }

    public static PlayerCharacter dummyPlayerCharacter() {
        Room fakeRoom = new Room("Fake Room", "For dummy character to exist in");
        return new PlayerCharacter("A Dummy Character", "To be used in the main menu", fakeRoom,
                                   Items.getCopiesOf(Items.BLUEBERRY, Items.FLOUR));
    }

    public void saveGame(Console console, PlayerCharacter playerCharacter) {
        try {
            new SaveGameSerial(playerCharacter.getName()
                                              .replace(' ', '_')
                                              .replace(File.separatorChar, '.'))
                    .saveData(playerCharacter);
            Printers.printMessage(console, "MainMenu.gameSaved");
        }
        catch (IOException | InterruptedException e1) {
            e1.printStackTrace();
        }
    }

    public MainMenu() {
        super(Settings.MESSAGES_BUNDLE.getString("MainMenu.name"), "");
    }

    public static class ExitToMainMenuException extends ExitToException {
    }

    @Override
    protected ActionList actions(GameCharacter actor, BaseGameEntity from) {
        ActionList actions = new ActionList();

        if (SaveGameSerial.hasActiveSave()) {
            SaveGameSerial activeSave = SaveGameSerial.loadActiveSave();
            if (activeSave != null && !(activeSave.loadData() instanceof DummyPlayerCharacter)) {
                // Only allow continue if the current player isn't a dummy
                actions.add(Actions.CONTINUE);
            }
        }

        actions.add(Actions.NEW_GAME);
        actions.add(Actions.EXIT_MENU);

        return actions;
    }

    @Override
    public int respondToInteraction(
            PlayerCharacter actor, BaseGameEntity from, Console console) throws ExitToException {
        Printers.println(console);
        Printers.print(console, this);
        Printers.println(console);

        BaseAction next = ConsolePrompt.getChoice(console, actions(actor, from), null);

        if (next.equals(Actions.CONTINUE)) {
            SaveGameSerial saveGame = SaveGameSerial.loadActiveSave();
            if (saveGame == null) {
                throw new IllegalStateException("SaveGame is null");
            }
            PlayerCharacter playerCharacter;
            playerCharacter = saveGame.loadData();
            return playerCharacter.interact(playerCharacter, this, console);
        }

        return next.apply(actor).interact(actor, this, console);
    }
}

File: src/bobbin/menus/MainMenu.kt
package bobbin.menus

import bobbin.boundaries.Room
import bobbin.characters.GameCharacter
import bobbin.characters.PlayerCharacter
import bobbin.constants.Actions
import bobbin.interaction.ConsolePrompt
import bobbin.interaction.ExitToException
import bobbin.interaction.Printers
import bobbin.interaction.actions.ActionList
import bobbin.interaction.console.Console
import bobbin.io.gamedata.SaveGameSerial
import bobbin.io.settings.Settings
import bobbin.items.BaseGameEntity
import java.io.File
import java.io.IOException
import java.util.*

open class MainMenu : Menu(Settings.MESSAGES_BUNDLE.getString("MainMenu.name"), "") {
    internal class DummyPlayerCharacter(name: String?, description: String?, location: Room?) : PlayerCharacter(name, description, location)

    @JvmOverloads
    fun saveGame(playerCharacter: PlayerCharacter, console: Console? = null) {
        Objects.requireNonNull(playerCharacter, "PlayerCharacter must be provided to save the game")
        try {
            SaveGameSerial(playerCharacter.name
                    .replace(' ', '_')
                    .replace(File.separatorChar, '.'))
                    .saveData(playerCharacter)
            if (console != null) {
                Printers.printMessage(console, "MainMenu.gameSaved")
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    class ExitToMainMenuException : ExitToException()

    override fun actions(actor: GameCharacter, from: BaseGameEntity): ActionList {
        val actions = ActionList()
        if (SaveGameSerial.hasActiveSave()) {
            val activeSave = SaveGameSerial.loadActiveSave()
            if (activeSave != null && activeSave.loadData() !is DummyPlayerCharacter) {
                // Only allow continue if the current player isn't a dummy
                actions.add(Actions.CONTINUE)
            }
        }
        actions.add(Actions.NEW_GAME)
        actions.add(Actions.EXIT_MENU)
        return actions
    }

    @Throws(ExitToException::class)
    override fun respondToInteraction(
            actor: PlayerCharacter, from: BaseGameEntity, console: Console): Int {
        currentPlayerCharacter = actor
        Printers.println(console)
        Printers.print(console, this)
        Printers.println(console)
        val next = ConsolePrompt.getChoice(console, actions(actor, from), null)
        if (next == Actions.CONTINUE) {
            val saveGame = SaveGameSerial.loadActiveSave() ?: throw IllegalStateException("SaveGame is null")
            val playerCharacter = saveGame.loadData()
            currentPlayerCharacter = playerCharacter
            return playerCharacter.interact(playerCharacter, this, console)
        }
        return next.apply(actor).interact(actor, this, console)
    }

    companion object {
        var currentPlayerCharacter: PlayerCharacter? = null
        fun dummyPlayerCharacter(): PlayerCharacter {
            val fakeRoom = Room("Fake Room", "For dummy character to exist in")
            return DummyPlayerCharacter("A Dummy Character", "To be used in the main menu", fakeRoom)
        }

        /**
         * Build a room, allowing the player to configure their character.
         *
         * @param console player input/output
         *
         * @return [PlayerCharacter] in the new game.
         */
        @JvmStatic
        fun buildNewGame(console: Console?): PlayerCharacter {
            val startingRoom = Room("Starting Room", "A Whole New Room")
            val playerCharacter = PlayerCharacter(
                    ConsolePrompt.getResponseString(console, "Character Name"),
                    ConsolePrompt.getResponseString(console, "Character Tagline"),
                    startingRoom)
            currentPlayerCharacter = playerCharacter
            return playerCharacter
        }
    }

    init {
        Runtime.getRuntime().addShutdownHook(Thread(Runnable {
            if (currentPlayerCharacter is PlayerCharacter && currentPlayerCharacter !is DummyPlayerCharacter) {
                saveGame(currentPlayerCharacter!!)
                println(Settings.MESSAGES_BUNDLE.getString("MainMenu.gameSaved"))
            }
        }))
    }
}